presto-main/src/main/java/com/facebook/presto/dispatcher/QueryInfoFetcher.java

```diff
@@ -0,0 +1,323 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.dispatcher;
+
+import com.facebook.airlift.concurrent.SetThreadName;
+import com.facebook.airlift.http.client.HttpClient;
+import com.facebook.airlift.http.client.Request;
+import com.facebook.airlift.http.client.ResponseHandler;
+import com.facebook.presto.execution.LocationFactory;
+import com.facebook.presto.execution.QueryState;
+import com.facebook.presto.execution.StateMachine;
+import com.facebook.presto.metadata.InternalNode;
+import com.facebook.presto.server.BasicQueryInfo;
+import com.facebook.presto.server.RequestErrorTracker;
+import com.facebook.presto.server.SimpleHttpResponseCallback;
+import com.facebook.presto.server.SimpleHttpResponseHandler;
+import com.facebook.presto.server.smile.BaseResponse;
+import com.facebook.presto.server.smile.Codec;
+import com.facebook.presto.server.smile.SmileCodec;
+import com.facebook.presto.spi.QueryId;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.Duration;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.ws.rs.core.UriBuilder;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.facebook.airlift.http.client.Request.Builder.prepareGet;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_CURRENT_STATE;
+import static com.facebook.presto.client.PrestoHeaders.PRESTO_MAX_WAIT;
+import static com.facebook.presto.server.RequestErrorTracker.queryRequestErrorTracker;
+import static com.facebook.presto.server.RequestHelpers.setContentTypeHeaders;
+import static com.facebook.presto.server.smile.AdaptingJsonResponseHandler.createAdaptingJsonResponseHandler;
+import static com.facebook.presto.server.smile.FullSmileResponseHandler.createFullSmileResponseHandler;
+import static com.facebook.presto.server.smile.JsonCodecWrapper.unwrapJsonCodec;
+import static com.facebook.presto.spi.StandardErrorCode.REMOTE_QUERY_ERROR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static io.airlift.units.Duration.nanosSince;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class QueryInfoFetcher
```

Comment:
Assume this class is forked from `TaskInfoFetcher`. Not sure if we can abstract it out a bit so we can reuse the common part. I do notice `SimpleHttpResponseCallback` has other implementations with similar implementations. Some existing code like `if(isBinaryTransportEnabled)/else` looks really like a bad design. Sad it's already there.

If it's too complicated, I'm ok leave this file as is.

Reply:
I added a TODO to clean up the binary transportation flag.  I don't think we can easily merge this entirely with the task infrastructure.